### PR TITLE
[OptionsResolver] Documenting nested options feature

### PR DESCRIPTION
**Pull request by yceruto**
Documents https://github.com/symfony/symfony/pull/27291

### CODE REVIEW

File: components/options_resolver.rst [hunk 1]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+            $resolver->setDefault('profiling', function (Options $options) {
+                return 'file' === $options['spool']['type'];
+            });
+        }
+    }
+
+.. note::
+
+    The fact that an option is defined as nested, means that you must to pass
+    an array of values to resolver it on runtime.
```
Comment (OskarStark) [inline components/options_resolver.rst]: s/resolver/resolve

Comment (yceruto) [reply components/options_resolver.rst]: Fixed, thanks.

File: components/options_resolver.rst [hunk 2]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
         }
     }
 
+Nested option
```
Comment (fabpot) [inline components/options_resolver.rst]: Nested Option

File: components/options_resolver.rst [hunk 3]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+~~~~~~~~~~~~~
+
+.. versionadded:: 4.2
+    This feature was introduced in Symfony 4.2.
+
+Suppose you want an option named ``spool`` which has two sub-options ``type``
+and ``path``. Instead of define it as a simple array of values, you can pass
```
Comment (fabpot) [inline components/options_resolver.rst]: defining

File: components/options_resolver.rst [hunk 4]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+a closure as the default value of the ``spool`` option with an :class:`Symfony\\Component\\OptionsResolver\\OptionsResolver`
```
Comment (fabpot) [inline components/options_resolver.rst]: with a

File: components/options_resolver.rst [hunk 5]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+            });
+        }
+    }
+
+.. caution::
+
+    The arguments of the closure must be type hinted as ``OptionsResolver`` and ``Options`` respectively.
+    Otherwise, the closure itself is considered as the default value of the option.
+
+In same way, parent options can access to the child option as follows::
```
Comment (fabpot) [inline components/options_resolver.rst]: can access the child

File: components/options_resolver.rst [hunk 6]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+            });
+            $resolver->setDefault('profiling', function (Options $options) {
+                return 'file' === $options['spool']['type'];
+            });
+        }
+    }
+
+.. note::
+
+    The fact that an option is defined as nested, means that you must to pass
```
Comment (fabpot) [inline components/options_resolver.rst]: no comma here.
you must pass
at runtime

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by yceruto** [comment]
@fabpot comments addressed, thanks!

**Comment by yceruto** [comment]
@javiereguiluz the code was merged. We can take care of this :) Do you have any suggestion? Thanks!

File: components/options_resolver.rst [hunk 7]
```diff
@@ -634,6 +634,105 @@ let you find out which options are defined::
...
+        // ...
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            // ...
+            $resolver->setDefault('spool', function (OptionsResolver $spoolResolver) {
+                $spoolResolver->setDefaults(array(
+                    'type' => 'file',
+                    // ...
+                ));
+                // ...
```
Comment (OskarStark) [inline components/options_resolver.rst]: So many of them 🙈🙈😊

Comment (yceruto) [reply components/options_resolver.rst]: I've removed some // ... :), thanks.

Comment (OskarStark) [reply components/options_resolver.rst]: thank you ;-)

**Review by OskarStark — APPROVED** [review]

**Comment by yceruto** [comment]
Please, don't forget this before final release of 4.2.

**Comment by yceruto** [comment]
Rebased on master.

**Comment by yceruto** [comment]
Someone could review this doc please. 4.2 is out there asking for this new feature :)

**Comment by javiereguiluz** [comment]
Thanks Yonel. I'm sorry it took us so long to merge this. I just did in 4.2 branch. Thanks!

**Comment by yceruto** [comment]
Thank you Javier!

**Pull request closed without merging**